ci/pip-base.sh

```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+source base.sh
+
+# Insert a custom compile command in generated requirements file so it is clear how they are generated:
+export CUSTOM_COMPILE_COMMAND="ci/pip-compile.sh"
+
+run_pip_compile() {
+    if [[ -f "requirements/requirements.txt" ]]; then
+        run pip-compile --output-file requirements/requirements.txt pyproject.toml
+    fi
+    run pip-compile --extra dev --output-file requirements/requirements-dev.txt pyproject.toml
+}
+
+run_pip_install() {
+    run pip install --ignore-installed --quiet --use-pep517 $@
```

Comment:
Why not `--upgrade` ?

Reply:
It was already like this, but I think we do it like this because we install all dependencies, direct and indirect, using an exact version spec, so `--upgrade` doesn't make sense.